packages/core/src/impl-store.ts

```diff
@@ -70,39 +83,119 @@ async function fetchOrDeployGeneric<T extends Deployment>(
+      if (isDevNet) {
+        debug('omitting a previous deployment due to no bytecode at address', stored.address);
+        stored = undefined;
+      } else {
+        throw new InvalidDeployment(stored, Reason.NoBytecode);
+      }
+    } else {
+      stored = validate(stored, deployment.getBytecodeHash(), hashBytecode(existingBytecode), isDevNet);
+    }
+  }
+  if (stored === undefined) {
+    deployment.set(undefined);
+  }
+  return stored;
+}
+
+function validate<T extends Deployment>(
+  deployment: T,
+  storedBytecodeHash: string | undefined,
+  existingBytecodeHash: string,
+  isDevNet: boolean,
+) {
+  if (storedBytecodeHash !== undefined && storedBytecodeHash !== existingBytecodeHash) {
+    if (isDevNet) {
+      debug('omitting a previous deployment due to mismatched bytecode at address ', deployment.address);
+      return undefined;
+    } else {
+      throw new InvalidDeployment(deployment, Reason.MismatchedBytecode);
+    }
+  }
+  return deployment;
+}
+
 export async function fetchOrDeploy(
   version: Version,
   provider: EthereumProvider,
   deploy: () => Promise<ImplDeployment>,
   opts?: DeployOpts,
+  merge?: boolean,
 ): Promise<string> {
-  return fetchOrDeployGeneric(implLens(version.linkedWithoutMetadata), provider, deploy, opts);
+  return fetchOrDeployGeneric(implLens(version.linkedWithoutMetadata), provider, deploy, opts, merge);
 }
 
 const implLens = (versionWithoutMetadata: string) =>
   lens(`implementation ${versionWithoutMetadata}`, 'implementation', data => ({
     get: () => data.impls[versionWithoutMetadata],
     set: (value?: ImplDeployment) => (data.impls[versionWithoutMetadata] = value),
+    getBytecodeHash: () => data.impls[versionWithoutMetadata]?.bytecodeHash,
+    merge: async (value?: ImplDeployment) => {
+      const existing = data.impls[versionWithoutMetadata];
+      if (existing !== undefined && value !== undefined) {
+        const { address, allAddresses } = await mergeAddresses(existing, value);
+        data.impls[versionWithoutMetadata] = { ...value, address, allAddresses };
+      } else {
+        data.impls[versionWithoutMetadata] = value;
+      }
+    },
   }));
 
+/**
+ * Merge the addresses in the deployments and returns them.
+ *
+ * @param existing existing deployment
+ * @param value deployment to add
+ */
+export async function mergeAddresses(existing: ImplDeployment, value: ImplDeployment) {
+  const merged = new Set<string>();
+
+  merged.add(existing.address);
+  merged.add(value.address);
+
+  if (existing.allAddresses !== undefined) {
+    existing.allAddresses.forEach(item => merged.add(item));
+  }
+  if (value.allAddresses !== undefined) {
+    value.allAddresses.forEach(item => merged.add(item));
+  }
+
+  return { address: existing.address, allAddresses: Array.from(merged) };
+}
+
 export async function fetchOrDeployAdmin(
   provider: EthereumProvider,
   deploy: () => Promise<Deployment>,
-  opts?: DeployOpts,
+  merge?: DeployOpts,
```

Comment:
Is it ever possible to merge an admin? Should this remain `opts`?

Reply:
Must have been a refactoring mistake.  It should be `opts`.